### PR TITLE
[Draft] Prototype approach for issue 3616

### PR DESCRIPTION
**Pull request by jlucovsky**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [3616](https://redmine.openinfosecfoundation.org/issues/3616)

Describe changes:
- Defer content body update until all transactions have been handled.

suricata-verify-pr: 348
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: src/detect.c [hunk 1]
```diff
@@ -1486,6 +1507,8 @@ static void DetectRunTx(ThreadVars *tv,
                     tx.tx_ptr, tx.tx_id,
                     flow_flags & STREAM_TOSERVER ? "toserver" : "toclient",
                     new_detect_flags);
+        if (alproto == ALPROTO_HTTP)
```
Comment (victorjulien) [inline src/detect.c:1510]: I would prefer a solution where we can avoid adding checks here.

One way would be to make sure `HttpServerBodyGetDataCallback` is still called only once, but we cache the output for the duration of the current run. `DetectEngineInspectBufferHttpHeader` does something similar for HTTP headers.

Comment (victorjulien) [reply src/detect.c:1510]: Though I have to say I don't really like the solution `http_header` uses either. But it does put the cost where it belongs, in the keyword implementation. This way its not a cost that all the other keywords and protocols also have.

Comment (jlucovsky) [reply src/detect.c]: @victorjulien This implements your second suggestion -- please review when you have time.

**Comment by catenacyber** [comment]
Is there a S-V test for this ?
Does it also fix https://redmine.openinfosecfoundation.org/issues/3691 ?

**Comment by victorjulien** [comment]
@jlucovsky do we have an SV test?

**Comment by jlucovsky** [comment]
> @jlucovsky do we have an SV test?

https://github.com/OISF/suricata-verify/pull/340

**Comment by jlucovsky** [comment]
> > @jlucovsky do we have an SV test?
> 
> [OISF/suricata-verify#340](https://github.com/OISF/suricata-verify/pull/340)

[OISF/suricata-verify#348](https://github.com/OISF/suricata-verify/pull/348)

**Comment by jlucovsky** [comment]
Continued in #5549

**Pull request closed without merging**